graphix/rng.py

```diff
@@ -0,0 +1,14 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+if TYPE_CHECKING:
+    from numpy.random import Generator
+
+
+def ensure_rng(rng: Generator | None = None) -> Generator:
```

Comment:
I believe rng is designed to be reused (i.e., typically instantiated only once), as its quality is argued for *different* draws from the *same* instance, rather than for different instances with distinct seeds.
Based on this observation, `ensure_rng` could be better implemented with thread-local storage (cache `rng`, maybe with weakref) ?

Reply:
I believe that when `ensure_rng` is called with a non-`None` argument, it should function as an identity. Usage scenarios can vary, and we should allow the caller to choose the implementation that best suits their needs. For example, it might be useful to run two simulations with distinct `rng` instances initialized with the same seed to compare them on the same measurement path.

As for the default behavior when `None` is passed, I am open to better suggestions. However, I think `get_default_rng()` should perform appropriately, and if it doesn't, we should consider proposing improvements to the implementation in `numpy` itself.

Reply:
> `get_default_rng()` should perform appropriately

Yeah.
May be a little bit slow though, as seed initialization requires true random number.
Let's discuss if it happened to be too slow.

Reply:
Sorry for the delay in understanding, but I now see your point. I've addressed it in commit 5cd6dee. Is this what you had in mind? Unfortunately, `np.random.Generator` does not support `weakref` yet.

Reply:
Exactly!
Let me submit some small fixes.

Reply:
How about something like this?

```python
from __future__ import annotations

import threading
import typing
from typing import TYPE_CHECKING

import numpy as np

if TYPE_CHECKING:
    from collections.abc import Callable

    from numpy.random import Generator


class _RNGStorage:
    rng: Generator | None


def _ensure_rng_wrapper() -> Callable[[Generator | None], Generator]:
    """Provide a hidden scope."""
    # HACK: Force mypy to check attributes
    rng_local = typing.cast(_RNGStorage, threading.local())
    rng_local.rng = None

    def ensure_rng(rng: Generator | None = None) -> Generator:
        if rng is not None:
            return rng
        if rng := rng_local.rng:
            return rng
        rng = np.random.default_rng()
        rng_local.rng = rng
        return rng

    return ensure_rng


ensure_rng = _ensure_rng_wrapper()
```

Reply:
In the line `if rng := rng_local.rng:` within `ensure_rng`, you cannot assume that the `rng` attribute exists on the thread-local object, because the initialization `rng_local.rng = None` is only performed for the current thread. This means that when `ensure_rng` is called in a different thread, the `rng` attribute may not be present, leading to an `AttributeError`. The new `test_rng.py` that I added in commit f694ad1 demonstrates this issue.

More generally, casting `threading.local()` to `_RNGStorage` is not type-safe, because the `rng` attribute does not always exist on instances of thread-local storage.

Regarding the introduction of `_ensure_rng_wrapper`, while it does hide `rng_local` from the module interface, it adds unnecessary complexity and introduces a new symbol (`_ensure_rng_wrapper`) into the module's scope. Although I understand the intent to protect `rng_local` from external modification, I don't believe this justifies the added complexity. If the goal is to prevent accidental mutation, simply renaming `rng_local` to `_rng_local` to indicate its internal use by convention should suffice. After all, if external code is determined to interfere with `ensure_rng`, it could do so regardless.

What do you think about this simpler approach?

Reply:
>  you cannot assume that the `rng` attribute exists on the thread-local object

Verified this behavior.
Thank you for helping me out!
Why not adding this kind of test to detect similar errors?

```python
def test_threadpool() -> None:
    with ThreadPoolExecutor() as executor:
        tasks = executor.map(lambda _: ensure_rng(None), range(100))
    for _ in tasks:
        pass
```
As for the code, I prefer `setattr` with noqa here to highlight the hidden risks.
I admit closure design is too much here...

```python
    rng = np.random.default_rng()
    # MEMO: Cannot perform type check
    setattr(_rng_local, "rng", rng)  # noqa: B010
```

Reply:
Fixed in 4d3d685 and 7fa5f00. Thanks!